doc/design/prefix_tree.md

```diff
@@ -0,0 +1,480 @@
+
+### **result = findPrefix(prefix, value)**
+
+Get the value pointers of all keys associated with a key with a given prefix.
+We also want information about the key of each of those values in contrast to exact queries like **find**
+these cannot be inferred from a prefix query.
+
+### Convenience functions
+
+1. get current size of the tree
+2. get all keys in the tree
+3. get all values in the tree
+4. get all keys with their corresponding values
+
+Other functions like his can be added as needed.
+
+## Solution
+
+The general strategy is to create a linked tree structure of specific nodes. There are two types of nodes,
+`DataNode`s containing user data and general `Node`s containing meta-data related to efficient search of the keys.
+
+Since we cannot use dynamic allocation, we use our own allocator specific for the task which is guaranteed to support the requested maximum number of values of the tree.
+
+### TypedAllocator
+
+This is a fast bucket allocator which allows allocation and deallocation of memory for a specific class type T
+as well as creation and destruction of objects of type T.
+The former corresponds to `malloc` and `free` and the latter to `new` and `delete`.
+
+Allocation and deallocation work in constant time, construction and destruction depend on compexity
+of the constructor and destructor.
+The allocator owns the memory pool it allocates from, is thread-safe, lock-free and relocatable.
+
+Internally it contains a lock-free queue and a buffer of memory blocks for objects of type T.
+The queue always contains indices to the free blocks (which can be allocated).
+
+If the allocator runs out of blocks, `allocate` or `create` will simply return `nullptr`
+while `deallocate` and `destroy` will never fail but must always be called with pointers previusly obtained by `allocate` or `create`.
+
+Two allocators are used to allocate nodes of one of the two node types and basically replace `new` and `delete`.
+The allocators need to be prepared for the worst-case of inserting a maximum number of of values each with keys sharing (almost) no prefixes.
+
+### Reloctable Pointer
+
+Internally we prefer pointers for computation and traversal whenever possible for efficiency.
+To ensure the data structure is relocatable, we have to use `relocatable_ptr` whenever we store internal pointers in the structure.
+This introduces no space overhead compared to raw pointers but a small additional runtime cost when dererencing the pointer.
+The benefit is that we can design the structure almost as if we were using raw pointers.
+
+Since the nodes need to contain pointers to establish the linked tree structure, we have to use `relocatable_ptr` there.
+We use the aliases
+
+* `node_ptr_t = relocatable_ptr<Node>`
+* `data_node_ptr_t = relocatable_ptr<DataNode>`
+
+for brevity.
+
+### Node
+
+We use a [de la briandais tree](https://www.scirp.org/%28S%28lz5mqp453edsnp55rrgjct55%29%29/reference/referencespapers.aspx?referenceid=1742945)
```

Comment:
I could not find find a good source for this, but this is the original paper. It is also not terribly important as it is concerned with the pointer structure of the nodes which I explain below.